src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface ILspServiceProvider
+{
+    ILspServices CreateServices(string serverKind);
+}
+
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+}
+
+public interface IRequestExecutionQueue<RequestContextType> where RequestContextType : struct
+{
+    Task<RequestContextType?> CreateRequestContextAsync(IQueueItem<RequestContextType> queueItem, CancellationToken cancellationToken);
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
```

Comment:
THis is from roslyn, but it might be worth having a short brainstorming session to see if we can avoid needing this events tangling pieces together.  Maybe you, me, DavidW